wps/services.py

```diff
@@ -0,0 +1,164 @@
+        params.update({'appid': self.api_key, 'units': 'metric'})
+        response = requests.get(self.url, params=params)
+        if (
+            response.status_code == 404
+            and response.json().get('message') == 'city not found'
+        ):
+            raise InvalidLocation(
+                f"Couldn't retrieve location weather. {response.text}"
+            )
+        if response.status_code != 200:
+            raise requests.exceptions.RequestException(response.text)
+        return response.json()
+
+    def get_location_temperature(
+        self, location: Union[Coordinates, CityInformation]
+    ):
+        if location.get('city'):
+            location = ','.join(
+                [
+                    location['city'].lower(),
+                    location['state'].lower(),
+                    location['country'].lower(),
+                ]
+            )
+            location = unidecode.unidecode(location)
+            params = {'q': location}
+        else:
+            params = {
+                'lat': location['latitude'],
+                'lon': location['longitude'],
+            }
+        weather_data = self.get_location_weather(params)
+        return weather_data.get('main').get('temp')
+
+
+def get_weather_api():
+    return OpenWeatherAPI()
+
+
+class PlaylistRecommendationAPI(abc.ABC):
+    @abc.abstractmethod
+    def get_recommendations(self, genres):
+        pass
+
+
+class SpotifyAPI(PlaylistRecommendationAPI):
+    def __init__(self):
+        self.headers = {'Authorization': f'Bearer {self._get_auth_token()}'}
+        self.recommendation_limit = 5
+        self.base_url = 'https://api.spotify.com/v1'
+
+    @staticmethod
+    def _get_auth_token():
+        client_id = settings.SPOTIFY_CLIENT_ID
+        client_secret = settings.SPOTIFY_CLIENT_SECRET
+        b64_client = base64.b64encode(
+            f'{client_id}:{client_secret}'.encode()
+        ).decode()
+        headers = {'Authorization': f'Basic {b64_client}'}
+        data = {'grant_type': 'client_credentials'}
+        url = 'https://accounts.spotify.com/api/token'
+        response = requests.post(url, data=data, headers=headers, timeout=30)
+        if response.status_code != 200:
+            raise requests.exceptions.RequestException(response.text)
+        access_token = response.json().get('access_token')
+        if not access_token:
+            raise requests.exceptions.RequestException(
+                f'Failed to retrieve access token, '
+                f'code: {response.status_code}, response: {response.text}'
+            )
+        return access_token
+
+    @staticmethod
+    def _format_recommendation_response(
+        response: dict,
+    ) -> List[Recommendation]:
+        if not response or not response.get('tracks'):
+            raise NoContentFound(
+                f'Invalid or missing content in response {response}'
+            )
+        formatted_data = []
```

Comment:
Could use list comprehension here

Reply:
I mean... sure, just think it will look kinda messy using list comprehension here. This seems far more readable